### PR TITLE
Add support for ServiceSplitter

### PR DESCRIPTION
**Pull request by ishustava**
Depends on #332 

Changes proposed in this PR:
- Add CRD support for the service splitter
- Update the version of controller-tools to 0.4.0 to support float types. With this version, we can pass `allowDangerousTypes` marker to allow CRDs to have `float32` types. This also comes with a breaking change to CRD and webhook versions, where now we have to explicitly set versions to `v1beta1` since controller-tools now defaults to `v1`.

How I've tested this PR:
Acceptance tests in hashicorp/consul-helm#624

How I expect reviewers to test this PR:
👀  acceptance tests in hashicorp/consul-helm#624

Checklist:
- [x] Tests added

### CODE REVIEW

File: api/v1alpha1/servicesplitter_types_test.go [hunk 1]
```diff
@@ -0,0 +1,317 @@
...
+					Name: "name",
+				},
+				Spec: ServiceSplitterSpec{},
+			},
+			Theirs: &capi.ServiceSplitterConfigEntry{
+				Kind:        capi.ServiceSplitter,
+				Name:        "name",
+				Namespace:   "namespace",
+				CreateIndex: 1,
+				ModifyIndex: 2,
```
Comment (ashwin-venkatesh) [inline api/v1alpha1/servicesplitter_types_test.go:30]: nice!

File: main.go [hunk 2]
```diff
@@ -4,8 +4,9 @@ import (
 	"log"
 	"os"
 
-	"github.com/hashicorp/consul-k8s/version"
 	"github.com/mitchellh/cli"
+
+	"github.com/hashicorp/consul-k8s/version"
```
Comment (ashwin-venkatesh) [inline main.go:9]: Probably revert this change?

File: subcommand/controller/command.go [hunk 3]
```diff
@@ -158,6 +155,15 @@ func (c *Command) Run(args []string) int {
 		setupLog.Error(err, "unable to create controller", "controller", common.ServiceRouter)
 		return 1
 	}
+	if err = (&controllers.ServiceSplitterController{
+		ConfigEntryController: configEntryReconciler,
+		Client:                mgr.GetClient(),
+		Log:                   ctrl.Log.WithName("controllers").WithName(common.ServiceSplitter),
```
Comment (ashwin-venkatesh) [inline subcommand/controller/command.go:161]:
```suggestion
		Log:                   ctrl.Log.WithName("controller").WithName(common.ServiceSplitter),
```

**Review by ashwin-venkatesh — COMMENTED** [review]
This looks phenomenal. Left a couple of suggestions but im still on the fence about the situation with `%` signs.

File: api/v1alpha1/servicesplitter_types.go [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+	var errs field.ErrorList
+	path := field.NewPath("spec").Child("splits")
+
+	// The sum of weights across all splits must add up to 100.
+	sumOfWeights := float32(0)
+	for i, split := range in.Spec.Splits {
+		// Validate that the weight value is between 0.01 and 100.
+		if split.Weight > 100 || split.Weight < 0.01 {
+			errs = append(errs, field.Invalid(path.Index(i).Child("weight"), split.Weight,
+				"weight must be between 0.01% and 100%"))
```
[the remark this answers is not in the document]
Comment (ishustava) [reply api/v1alpha1/servicesplitter_types.go:155]: yeah, that's a good point. I like your suggestion.

File: api/v1alpha1/servicesplitter_types.go [hunk 5]
```diff
@@ -0,0 +1,191 @@
...
+	return reflect.DeepEqual(in.ToConsul(), candidate)
+}
+
+func (in *ServiceSplitter) Validate() error {
+	var errs field.ErrorList
+	path := field.NewPath("spec").Child("splits")
+
+	// The sum of weights across all splits must add up to 100.
+	sumOfWeights := float32(0)
+	for i, split := range in.Spec.Splits {
```
[the remark this answers is not in the document]
Comment (ishustava) [reply api/v1alpha1/servicesplitter_types.go:151]: 🖖 absolutely!

File: api/v1alpha1/servicesplitter_types_test.go (hunk 1 above)
Comment (ishustava) [reply api/v1alpha1/servicesplitter_types_test.go:30]: thank you! I pride myself in being good at copy-pasting code 😆

File: api/v1alpha1/servicesplitter_types.go [hunk 6]
```diff
@@ -0,0 +1,191 @@
...
+				"weight must be between 0.01% and 100%"))
+		}
+
+		// If valid, add to sumOfWeights.
+		sumOfWeights += split.Weight
+	}
+
+	if sumOfWeights != 100 {
+		errs = append(errs, field.Invalid(path, sumOfWeights,
+			"the sum of weights across all splits must add up to 100%"))
```
[the remark this answers is not in the document]
Comment (ishustava) [reply api/v1alpha1/servicesplitter_types.go:164]: yeah the invalid field is all the splits. The error message looks like this currently:
`servicesplitter.consul.hashicorp.com "foo" is invalid: spec.splits: Invalid value: 95: the sum of weights across all splits must add up to 100%`

> I'd expect the invalid field to be the json encoded array of splits.

I'm not sure what you mean by that 🙂

File: api/v1alpha1/servicesplitter_types.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (ishustava) [reply api/v1alpha1/servicesplitter_types.go:155]: yeah, will rephrase like Ashwin suggested above.

File: controllers/configentry_controller_test.go [hunk 7]
```diff
@@ -182,9 +222,11 @@ func TestConfigEntryControllers_createsConfigEntry(t *testing.T) {
 			})
 			req.NoError(err)
 			if c.consulPrereq != nil {
```
[the remark this answers is not in the document]
Comment (ishustava) [reply controllers/configentry_controller_test.go:224]: Good catch! will do that!

File: main.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (ishustava) [reply main.go:9]: I'll re-order it back to be like on the master branch. don't know why and how this happened.

File: api/v1alpha1/servicesplitter_types.go [hunk 8]
```diff
@@ -0,0 +1,228 @@
...
+		}
+
+		// If valid, add to sumOfWeights.
+		sumOfWeights += split.Weight
+	}
+
+	if sumOfWeights != 100 {
+		errs = append(errs, field.Invalid(path, sumOfWeights,
+			"the sum of weights across all splits must add up to a 100 percent"))
+	}
```
Comment (ashwin-venkatesh) [inline api/v1alpha1/servicesplitter_types.go:165]:
```suggestion
	errs := in.Spec.Splits.validate()
```

File: api/v1alpha1/servicesplitter_types.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (ashwin-venkatesh) [reply api/v1alpha1/servicesplitter_types.go]: I think he means to replace `sumOfWeights` in the error message with the json encode `splits` slice

[the remark this answers is not in the document]
Comment (ishustava) [reply api/v1alpha1/servicesplitter_types.go]: Ah I see, yeah that would be much better! I'll update it.

**Review by ashwin-venkatesh — APPROVED** [review]
🏅

**Review by lkysow — APPROVED** [review]
Awesome!
